### ISSUE TITLE
Unable to get Rsyslog structured data in syslog message

### ISSUE THREAD

**Issue by RanjodhKaur** [Graylog2/graylog2-server#1084]
I am new to syslog module. I am trying to log messages with structured data by using this parameter in template. But it is showing null value for structured data. I am working with rsyslogd 5.8.11 and rsyslogd 4.8.4 and for both the versions i am getting same results.

I am using the template:

"<%PRI%>%TIMESTAMP:::date-rfc3339%%HOSTNAME%%syslogtag%%APP-NAME%%PROCID%%MSGID% %msg% %STRUCTURED-DATA%\n”

getting message format as below:
<142>  2015-01-29T06:43:53.081641-05:00 localhost login[2116]: login 2116 -  [2116 : 2116 INFO]SERIAL Login from IP:127.0.0.1 user:admin -

can someone tell me what I am doing wrong?.I just want to log message with structured data with SD-ID,SD-PARAM in log file.

**Comment by joschi** [comment]
Please try again with the message template described in http://docs.graylog.org/en/1.0/pages/sending_data.html#rsyslog.

**Comment by LilyChadha** [comment]
Tried with suggested template.Still getting same thing.Can you suggest something else?

**Comment by fwdchrisull** [comment]
I have encountered a lot of weird issues working with older version of rsyslogd.  I would recommend updating to a newer version if possible.  I didn't have any problems with this template in v7.

**Comment by shakalandy** [comment]
Same over here with rsyslog-7.4.7-7.el7_0.x86_64.
Template as described on graylog docs:
```
$template GRAYLOGRFC5424,"<%PRI%>%PROTOCOL-VERSION% %TIMESTAMP:::date-rfc3339% %HOSTNAME% %APP-NAME% %PROCID% %MSGID% %STRUCTURED-DATA% %msg%\n"
*.* @myloghost:12201;GRAYLOGRFC5424
```
Graylog is waiting for syslog-udp, "Expand structured data" disabled.

The Message field contains 2 leading dashs after indexing:
```
- -  [origin software="rsyslogd" swVersion="7.4.7" x-pid="12122" x-info="http://www.rsyslog.com"] start
```

What's wrong?
Graylog Version is latest.

**Comment by fwdchrisull** [comment]
I've got the leading dashes as well, but since all the other indexed fields were populating so I was willing to live with this.  From what I've found on other forums, the dashes are there due to nil values in structured data fields due to conversion from rfc3164 to 5424.  I'm not sure where to go from here so maybe an rsyslog expert can provide more info.

**Comment by LilyChadha** [comment]
@fwdchrisull   you are right leading two dashes are cause of nil values only.But reason behind this is not the conversion from rfc3164 to 5424. These NIL VALUES repesents that given parameter is empty.

@shakalandy   in your template it is showing nil value for MSG-ID and STRUCTURED DATA.

I am also getting NIL VALUE for STRUCTURED DATA.

**Comment by fwdchrisull** [comment]
Can you post an example raw message?

**Comment by shakalandy** [comment]
for the above mentioned "message"
```
- -  [origin software="rsyslogd" swVersion="7.4.7" x-pid="16675" x-info="http://www.rsyslog.com"] start
```
the full_message looks like this:

```
<46>0 2015-04-07T17:17:40.855688+02:00 hostname rsyslogd - - -  [origin software="rsyslogd" swVersion="7.4.7" x-pid="16675" x-info="http://www.rsyslog.com"] start
```

**Comment by LilyChadha** [comment]
For me the message thrown by syslogd daemon has below format:
<PRI >=<46> i.e 5*8+6   ----------------> 5 is for messages logged by syslogd daemon  6 for INFO messages

<46>  2015-01-01T00:15:03.808718-05:00 localhost rsyslogd: rsyslogd - -  [origin software="rsyslogd" swVersion="4.6.4" x-pid="1164" x-info="http://www.rsyslog.com"] rsyslogd was HUPed, type 'lightweight'. -

here i am getting structured data in my message.But if i am using syslog call with local1 i am not getting structured data:

<PRI >=<142> i.e 17*8+6 ---------------->17 is for local1..6 for INFO messages
<142> 2015-01-29T06:43:53.081641-05:00 localhost login[2116]: login 2116 - [2116 : 2116 INFO]SERIAL Login from IP:127.0.0.1 user:admin -

template used:
<%PRI%>%TIMESTAMP:::date-rfc3339%%HOSTNAME%%syslogtag%%APP-NAME%%PROCID%%MSGID% %msg% %STRUCTURED-DATA%\n”

**Comment by bernd** [comment]
If you want to send RFC5424 syslog to Graylog, you have to configure the correct template for rsyslog as documented [here](http://docs.graylog.org/en/1.0/pages/sending_data.html#rsyslog). (or use the predefined `RSYSLOG_SyslogProtocol23Format` template)

The `%MSGID%` and `%STRUCTURED-DATA%` can be `NIL` (`-`) if there is no structured data and/or no message id. For regular syslog they will mostly be `NIL`.

The problem with the leading `- - ` in the parsed message is due to a problem in our syslog parser. It did not strip `NIL` values for the message id and structured data. This is addressed in #1161 and will be fixed in the upcoming 1.1 release of Graylog.

I am closing this issue. Please re-open if you still have problems after the 1.1 release. Thank you!
